// edge-lexer/build/src/utils.js
import { n as TagTypes, t as MustacheTypes } from "../enums-Bj4v9G-x.js";
function isTag(token, name) {
	if (token.type === TagTypes.TAG || token.type === TagTypes.ETAG) return name ? token.properties.name === name : true;
	return false;
}
function isEscapedTag(token, name) {
	if (token.type === TagTypes.ETAG) return name ? token.properties.name === name : true;
	return false;
}
function isMustache(token) {
	return token.type === MustacheTypes.EMUSTACHE || token.type === MustacheTypes.ESMUSTACHE || token.type === MustacheTypes.MUSTACHE || token.type === MustacheTypes.SMUSTACHE;
}
function isSafeMustache(token) {
	return token.type === MustacheTypes.ESMUSTACHE || token.type === MustacheTypes.SMUSTACHE;
}
function isEscapedMustache(token) {
	return token.type === MustacheTypes.EMUSTACHE || token.type === MustacheTypes.ESMUSTACHE;
}
function getLineAndColumn(token) {
	if (token.type === "newline" || token.type === "raw") return [token.line, 0];
	return [token.loc.start.line, token.loc.start.col];
}
export { getLineAndColumn, isEscapedMustache, isEscapedTag, isMustache, isSafeMustache, isTag };
