// edge-parser/build/index.js
import "node:module";
import { EOL } from "node:os";
import { MustacheTypes, TagTypes, Tokenizer } from "edge-lexer";
import { generate } from "astring";
import { EdgeError } from "edge-error";
import { parse } from "acorn";
import stringify from "js-stringify";
var __defProp = Object.defineProperty;
var __export = (all, symbols) => {
	let target = {};
	for (var name in all) __defProp(target, name, {
		get: all[name],
		enumerable: true
	});
	if (symbols) __defProp(target, Symbol.toStringTag, { value: "Module" });
	return target;
};
var Stack = class {
	#localVariables = [];
	#scopes = [];
	#getRecentScope() {
		return this.#scopes.length ? this.#scopes[this.#scopes.length - 1] : this.#localVariables;
	}
	#isInList(list, item) {
		return !!list.find((listItem) => listItem === item);
	}
	defineScope() {
		this.#scopes.push([]);
	}
	clearScope() {
		this.#scopes.pop();
	}
	defineVariable(variableName) {
		this.#getRecentScope().push(variableName);
	}
	has(variableName) {
		if (this.#isInList(this.#localVariables, variableName)) return true;
		return !!this.#scopes.find((scope) => this.#isInList(scope, variableName));
	}
	getState() {
		return {
			localVariables: this.#localVariables,
			scopes: this.#scopes
		};
	}
	list() {
		return this.#scopes.flat().concat(this.#localVariables);
	}
};
function stringify$1(astExpression) {
	return generate(astExpression);
}
function patchLoc(loc, lexerLoc) {
	if (loc.start.line === 1) loc.start.column = loc.start.column + lexerLoc.start.col;
	loc.start.line = loc.start.line + lexerLoc.start.line - 1;
	loc.end.line = loc.end.line + lexerLoc.start.line - 1;
}
function generateAST(jsArg, lexerLoc, filename) {
	const acornOptions = {
		locations: true,
		ecmaVersion: 2020,
		allowAwaitOutsideFunction: true,
		onToken: (token) => patchLoc(token.loc, lexerLoc)
	};
	try {
		return parse(jsArg, acornOptions)["body"][0];
	} catch (error) {
		const line = error.loc.line + lexerLoc.start.line - 1;
		const col = error.loc.line === 1 ? error.loc.column + lexerLoc.start.col : error.loc.column;
		throw new EdgeError(error.message.replace(/\(\d+:\d+\)/, ""), "E_ACORN_ERROR", {
			line,
			col,
			filename
		});
	}
}
function makeMemberAccessor(propertyName, args) {
	return {
		type: "MemberExpression",
		object: {
			type: "Identifier",
			name: propertyName
		},
		computed: false,
		property: args
	};
}
var identifier_default = { toStatement(statement, _, parser) {
	if ((parser.options.localVariables || []).indexOf(statement.name) > -1 || parser.stack.has(statement.name) || global[statement.name] !== void 0) return statement;
	return makeMemberAccessor(parser.options.statePropertyName, statement);
} };
var member_expression_default = { toStatement(statement, filename, parser) {
	statement.object = transformAst(statement.object, filename, parser);
	if (statement.computed || statement.property.type !== "Identifier") statement.property = transformAst(statement.property, filename, parser);
	return statement;
} };
var expression_statement_default = { toStatement(statement, filename, parser) {
	return transformAst(statement.expression, filename, parser);
} };
var call_expression_default = { toStatement(statement, filename, parser) {
	statement.callee = transformAst(statement.callee, filename, parser);
	statement.arguments = statement.arguments.map((node) => transformAst(node, filename, parser));
	return statement;
} };
var arrow_function_expression_default = { toStatement(statement, filename, parser) {
	parser.stack.defineScope();
	statement.params.forEach((param) => {
		if (param.type === "Identifier") parser.stack.defineVariable(param.name);
		else if (param.type === "ObjectPattern") parser.utils.collectObjectExpressionProperties(param).forEach((prop) => {
			parser.stack.defineVariable(prop);
		});
		else if (param.type === "ArrayPattern") parser.utils.collectArrayExpressionProperties(param).forEach((prop) => {
			parser.stack.defineVariable(prop);
		});
		else {
			const { line, col } = parser.utils.getExpressionLoc(param);
			throw new EdgeError(`Report this error to the maintainers: Unexpected arrow function property type ${param.type}`, "E_PARSER_ERROR", {
				line,
				col,
				filename
			});
		}
	});
	statement.body = transformAst(statement.body, filename, parser);
	parser.stack.clearScope();
	return statement;
} };
var literal_default = { toStatement(statement) {
	return statement;
} };
var template_literal_default = { toStatement(statement, filename, parser) {
	statement.expressions = statement.expressions.map((expression) => {
		return transformAst(expression, filename, parser);
	});
	return statement;
} };
var binary_expression_default = { toStatement(statement, filename, parser) {
	statement.left = transformAst(statement.left, filename, parser);
	statement.right = transformAst(statement.right, filename, parser);
	return statement;
} };
var array_expression_default = { toStatement(statement, filename, parser) {
	statement.elements = statement.elements.map((element) => transformAst(element, filename, parser));
	return statement;
} };
var object_expression_default = { toStatement(statement, filename, parser) {
	statement.properties = statement.properties.map((node) => {
		if (node.type === "Property") {
			node.shorthand = false;
			if (node.computed === true) node.key = transformAst(node.key, filename, parser);
			node.value = transformAst(node.value, filename, parser);
			return node;
		}
		if (node.type === "SpreadElement") return transformAst(node, filename, parser);
		const { line, col } = parser.utils.getExpressionLoc(node);
		throw new EdgeError(`Report this error to the maintainers: Unexpected object property type "${node.type}"`, "E_PARSER_ERROR", {
			line,
			col,
			filename
		});
	});
	return statement;
} };
var unary_expression_default = { toStatement(statement, filename, parser) {
	statement.argument = transformAst(statement.argument, filename, parser);
	return statement;
} };
var function_declaration_default = { toStatement(statement) {
	return statement;
} };
var conditional_expression_default = { toStatement(statement, filename, parser) {
	statement.test = transformAst(statement.test, filename, parser);
	statement.consequent = transformAst(statement.consequent, filename, parser);
	statement.alternate = transformAst(statement.alternate, filename, parser);
	return statement;
} };
var logical_expression_default = { toStatement(statement, filename, parser) {
	statement.left = transformAst(statement.left, filename, parser);
	statement.right = transformAst(statement.right, filename, parser);
	return statement;
} };
var sequence_expression_default = { toStatement(statement, filename, parser) {
	statement.expressions = statement.expressions.map((expression) => {
		return transformAst(expression, filename, parser);
	});
	return statement;
} };
var assignment_expression_default = { toStatement(statement, filename, parser) {
	statement.left = transformAst(statement.left, filename, parser);
	statement.right = transformAst(statement.right, filename, parser);
	return statement;
} };
const UNALLOWED_EXPRESSION_MESSAGE = "Make sure to render template in async mode before using await expression";
var await_expression_default = { toStatement(statement, filename, parser) {
	if (!parser.options.async) {
		const { line, col } = parser.utils.getExpressionLoc(statement);
		throw new EdgeError(UNALLOWED_EXPRESSION_MESSAGE, "E_PARSER_ERROR", {
			line,
			col,
			filename
		});
	}
	statement.argument = transformAst(statement.argument, filename, parser);
	return statement;
} };
var new_expression_default = { toStatement(statement, filename, parser) {
	statement.arguments = statement.arguments.map((expression) => {
		return transformAst(expression, filename, parser);
	});
	return statement;
} };
var block_statement_default = { toStatement(statement, filename, parser) {
	statement.body = statement.body.map((token) => {
		return transformAst(token, filename, parser);
	});
	return statement;
} };
var return_statement_default = { toStatement(statement, filename, parser) {
	statement.argument = transformAst(statement.argument, filename, parser);
	return statement;
} };
var this_expression_default = { toStatement(statement) {
	return statement;
} };
var chain_expression_default = { toStatement(statement, filename, parser) {
	statement.expression = transformAst(statement.expression, filename, parser);
	return statement;
} };
var spread_element_default = { toStatement(statement, filename, parser) {
	statement.argument = parser.utils.transformAst(statement.argument, filename, parser);
	return statement;
} };
var update_expression_default = { toStatement(statement, filename, parser) {
	statement.argument = transformAst(statement.argument, filename, parser);
	return statement;
} };
var expressions_exports = /* @__PURE__ */ __export({
	ArrayExpression: () => array_expression_default,
	ArrowFunctionExpression: () => arrow_function_expression_default,
	AssignmentExpression: () => assignment_expression_default,
	AwaitExpression: () => await_expression_default,
	BinaryExpression: () => binary_expression_default,
	BlockStatement: () => block_statement_default,
	CallExpression: () => call_expression_default,
	ChainExpression: () => chain_expression_default,
	ConditionalExpression: () => conditional_expression_default,
	ExpressionStatement: () => expression_statement_default,
	FunctionDeclaration: () => function_declaration_default,
	Identifier: () => identifier_default,
	Literal: () => literal_default,
	LogicalExpression: () => logical_expression_default,
	MemberExpression: () => member_expression_default,
	NewExpression: () => new_expression_default,
	ObjectExpression: () => object_expression_default,
	ReturnStatement: () => return_statement_default,
	SequenceExpression: () => sequence_expression_default,
	SpreadElement: () => spread_element_default,
	TemplateLiteral: () => template_literal_default,
	ThisExpression: () => this_expression_default,
	UnaryExpression: () => unary_expression_default,
	UpdateExpression: () => update_expression_default
});
function transformAst(astExpression, filename, parser) {
	const Expression = expressions_exports[astExpression.type];
	if (Expression) return Expression.toStatement(astExpression, filename, parser);
	const { type, loc } = astExpression;
	throw new EdgeError(`"${type}" is not supported`, "E_UNALLOWED_EXPRESSION", {
		line: loc.start.line,
		col: loc.start.column,
		filename
	});
}
function makeCallable(paths, args) {
	if (typeof paths === "string") return {
		type: "CallExpression",
		callee: {
			type: "Identifier",
			name: paths
		},
		arguments: args
	};
	return {
		type: "CallExpression",
		callee: {
			type: "MemberExpression",
			object: {
				type: "Identifier",
				name: paths[0]
			},
			property: {
				type: "Identifier",
				name: paths[1]
			}
		},
		arguments: args
	};
}
function collectObjectExpressionProperties(expression) {
	return expression.properties.map((prop) => {
		if (prop.value.type !== "Identifier") throw new Error("Object destructuring should not reference dynamic properties");
		return prop.value.name;
	});
}
function collectArrayExpressionProperties(expression) {
	return expression.elements.map((prop) => {
		if (prop.type !== "Identifier") throw new Error("Array destructuring should not reference dynamic properties");
		return prop.name;
	});
}
var Parser = class {
	asyncMode;
	constructor(tags, stack = new Stack(), options) {
		this.tags = tags;
		this.stack = stack;
		this.options = options;
		this.asyncMode = !!this.options.async;
	}
	utils = {
		stringify: stringify$1,
		transformAst,
		makeCallable,
		makeMemberAccessor,
		generateAST,
		makeEscapeCallable: makeCallable,
		makeStatePropertyAccessor: makeMemberAccessor,
		collectObjectExpressionProperties,
		collectArrayExpressionProperties,
		getExpressionLoc(expression) {
			const loc = expression.loc || expression.property?.loc;
			return {
				line: loc.start.line,
				col: loc.start.column
			};
		}
	};
	#getTokenizerOptions(options) {
		if (!this.options) return options;
		return {
			claimTag: this.options.claimTag,
			onLine: this.options.onLine,
			filename: options.filename
		};
	}
	#processEscapedTagToken(token, buffer) {
		`@${token.properties.name}(${token.properties.jsArg})`.split("\n").forEach((line) => buffer.outputRaw(line));
		token.children.forEach((child) => this.processToken(child, buffer));
		buffer.outputRaw(`@end${token.properties.name}`);
	}
	#processEscapedMustache(token, buffer) {
		(token.type === MustacheTypes.EMUSTACHE ? `{{${token.properties.jsArg}}}`.split("\n") : `{{{${token.properties.jsArg}}}}`.split("\n")).forEach((line) => buffer.outputRaw(line));
	}
	#processMustache({ properties, loc, filename, type }, buffer) {
		const node = transformAst(generateAST(properties.jsArg, loc, filename), filename, this);
		const expression = type === MustacheTypes.MUSTACHE ? makeCallable(this.options.escapeCallPath, [node]) : node;
		if (node.type === "TemplateLiteral") buffer.outputExpression(stringify$1(expression), filename, loc.start.line, false);
		else if (node.type === "FunctionDeclaration") buffer.outputExpression(stringify$1(node), filename, loc.start.line, false);
		else buffer.outputExpression(stringify$1(expression), filename, loc.start.line, true);
	}
	tokenize(template, options) {
		const tokenizer = new Tokenizer(template, this.tags, this.#getTokenizerOptions(options));
		tokenizer.parse();
		return tokenizer.tokens;
	}
	processToken(token, buffer) {
		switch (token.type) {
			case "raw":
				buffer.outputRaw(token.value);
				break;
			case "newline":
				buffer.outputRaw(EOL === "\n" ? "\n" : "\r\n");
				break;
			case TagTypes.TAG:
				if (typeof this.options.onTag === "function") this.options.onTag(token);
				this.tags[token.properties.name].compile(this, buffer, token);
				break;
			case TagTypes.ETAG:
				this.#processEscapedTagToken(token, buffer);
				break;
			case MustacheTypes.EMUSTACHE:
			case MustacheTypes.ESMUSTACHE:
				this.#processEscapedMustache(token, buffer);
				break;
			case MustacheTypes.SMUSTACHE:
			case MustacheTypes.MUSTACHE:
				if (typeof this.options.onMustache === "function") this.options.onMustache(token);
				this.#processMustache(token, buffer);
		}
	}
};
var EdgeBuffer = class EdgeBuffer {
	#outputFileAndLineNumber = true;
	#outputOutVariable = true;
	#outputReturnStatement = true;
	#wrapInsideTryCatch = true;
	#options = {
		outputVar: "",
		rethrowCallPath: "",
		fileNameVar: "$filename",
		lineVar: "$lineNumber"
	};
	#prefix = [];
	#suffix = [];
	#buffer = [];
	#currentLineNumber = 1;
	#filename;
	#currentFileName;
	#compiledOutput;
	outputVariableName;
	constructor(filename, options) {
		this.#filename = filename;
		this.#currentFileName = this.#filename;
		this.outputVariableName = options.outputVar;
		this.#options.outputVar = options.outputVar;
		this.#options.rethrowCallPath = Array.isArray(options.rethrowCallPath) ? options.rethrowCallPath.join(".") : options.rethrowCallPath;
	}
	create(filename, options) {
		return new EdgeBuffer(filename, Object.assign({}, this.#options, options));
	}
	get size() {
		return this.#buffer.length;
	}
	#setup(buffer) {
		this.#outputOutVariable && buffer.push(`let ${this.outputVariableName} = "";`);
		this.#outputFileAndLineNumber && buffer.push(`let ${this.#options.lineVar} = 1;`);
		this.#outputFileAndLineNumber && buffer.push(`let ${this.#options.fileNameVar} = ${stringify(this.#filename)};`);
		this.#wrapInsideTryCatch && buffer.push("try {");
	}
	#teardown(buffer) {
		if (this.#wrapInsideTryCatch) {
			buffer.push("} catch (error) {");
			buffer.push(`${this.#options.rethrowCallPath}(error, ${this.#options.fileNameVar}, ${this.#options.lineVar});`);
			buffer.push("}");
		}
		this.#outputReturnStatement && buffer.push(`return ${this.outputVariableName};`);
	}
	#updateFileName(filename) {
		if (this.#currentFileName !== filename) {
			this.#currentFileName = filename;
			this.#buffer.push(`${this.#options.fileNameVar} = ${stringify(filename)};`);
		}
	}
	#updateLineNumber(lineNumber) {
		if (lineNumber > 0 && this.#currentLineNumber !== lineNumber) {
			this.#currentLineNumber = lineNumber;
			this.#buffer.push(`${this.#options.lineVar} = ${lineNumber};`);
		}
	}
	outputRaw(text) {
		this.#buffer.push(`${this.outputVariableName} += ${stringify(text)};`);
		return this;
	}
	outputExpression(text, filename, lineNumber, templateLiteral) {
		this.#updateFileName(filename);
		this.#updateLineNumber(lineNumber);
		text = templateLiteral ? `\`\${${text}}\`` : text;
		this.#buffer.push(`${this.outputVariableName} += ${text};`);
		return this;
	}
	writeExpression(text, filename, lineNumber) {
		this.#updateFileName(filename);
		this.#updateLineNumber(lineNumber);
		this.#buffer.push(`${text};`);
		return this;
	}
	writeStatement(text, filename, lineNumber) {
		this.#updateFileName(filename);
		this.#updateLineNumber(lineNumber);
		this.#buffer.push(`${text}`);
		return this;
	}
	wrap(prefix, suffix) {
		this.#prefix.push(prefix);
		this.#suffix.push(suffix);
		return this;
	}
	disableFileAndLineVariables() {
		this.#outputFileAndLineNumber = false;
		return this;
	}
	disableOutVariable() {
		this.#outputOutVariable = false;
		return this;
	}
	disableReturnStatement() {
		this.#outputReturnStatement = false;
		return this;
	}
	disableTryCatchBlock() {
		this.#wrapInsideTryCatch = false;
		return this;
	}
	flush() {
		if (this.#compiledOutput !== void 0) return this.#compiledOutput;
		let buffer = [];
		this.#prefix.forEach((text) => text.split(EOL).forEach((line) => buffer.push(`${line}`)));
		this.#setup(buffer);
		buffer = buffer.concat(this.#buffer);
		this.#teardown(buffer);
		this.#suffix.forEach((text) => text.split(EOL).forEach((line) => buffer.push(`${line}`)));
		this.#compiledOutput = buffer.join(EOL);
		return this.#compiledOutput;
	}
};
const expressions = Object.keys(expressions_exports).reduce((result, name) => {
	result[name] = name;
	return result;
}, {});
export { EdgeBuffer, Parser, Stack, expressions };
